From 5f2ddccaa89b21b66d7d9c6dbba71ce3937ff09f Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Mon, 27 Feb 2023 14:12:23 -0500
Subject: [PATCH] cgroup: Stumble on if we can't set up oom handling

OpenShift 4.13 tentatively plans to use cgroupv1 on a RHEL9
era kernel, which breaks only in `PREEMPT_RT` aka `kernel-rt`
due to
https://github.com/torvalds/linux/commit/2343e88d238f5de973d609d861c505890f94f22e

For now let's stumble on without oom notifications.

Signed-off-by: Colin Walters <walters@verbum.org>

Upstream-Status: Backport [https://github.com/containers/conmon/commit/5cff0cea135254b99e1a99d7566848f1f6755387]
Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 src/cgroup.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/cgroup.c b/src/cgroup.c
index 48f611a..19c08fb 100644
--- a/src/cgroup.c
+++ b/src/cgroup.c
@@ -157,8 +157,15 @@ static void setup_oom_handling_cgroup_v1(int pid)
 		pexit("Failed to create eventfd");
 
 	_cleanup_free_ char *data = g_strdup_printf("%d %d", oom_event_fd, oom_cgroup_fd);
-	if (write_all(cfd, data, strlen(data)) < 0)
-		pexit("Failed to write to cgroup.event_control");
+	if (write_all(cfd, data, strlen(data)) < 0) {
+		/* This used to be fatal, but we make it advisory and stumble on because
+		 * https://github.com/torvalds/linux/commit/2343e88d238f5de973d609d861c505890f94f22e
+		 * disables this interface in PREEMPT_RT kernel configs.
+		 */
+		nwarnf("Failed to write to cgroup.event_control");
+		g_free(memory_cgroup_file_path);
+		return;
+	}
 
 	g_unix_fd_add(oom_event_fd, G_IO_IN, oom_cb_cgroup_v1, memory_cgroup_file_path);
 }
-- 
2.27.0

